### PR TITLE
Partest supports Crash?! again and X failure in terse mode

### PR DESCRIPTION
**Pull request by som-snytt**
More partest janitorial duty, inspired by SZ.

Notably, steps or actions in a test produce `TestState` instead of boolean, to support `Crash` status. Previously, a compiler crash resulted in a generic failure.

A crash `?!` result was witnessed manually by injecting a crasher; outright crashes are so rare now that I didn't go looking for one.

### CODE REVIEW

File: src/partest/scala/tools/partest/nest/AbstractRunner.scala [hunk 1]
```diff
@@ -73,26 +73,32 @@ class AbstractRunner(val config: RunnerSpec.Config, protected final val testSour
...
+      val logged = if (logOnFail) showLog() else Nil
+      diffed ::: logged
+    }
+    if (terse) {
+      if (state.isOk) { printDot() ; Nil }
+      else { printEx() ; statusLine(state, durationMs) :: errInfo }
+    } else {
+      echo(statusLine(state, durationMs))
+      if (!state.isOk) errInfo.foreach(echo)
+      Nil
```
Comment (szeiger) [inline src/partest/scala/tools/partest/nest/AbstractRunner.scala:101]: Is this supposed to return `Nil` when there were errors?

Comment (som-snytt) [reply src/partest/scala/tools/partest/nest/AbstractRunner.scala:101]: IIRC this is just a way to smuggle terse-mode error output to be printed later. If not terse, the output is immediate. I also remember not liking the hack at L313. So this interaction could be encapsulated better.

**Review by szeiger — APPROVED** [review]

**Comment by som-snytt** [comment]
Just to show how charming the terse output looks. Also tidy. The X is in red.

```
# starting 1027 tests in neg
.........X..............................................................
........................................................................
........................................................................
........................................................................
.........................................................................
........................................................................
.........................................................................
.........................................................................
........................................................................
........................................................................
........................................................................
........................................................................
........................................................................
........................................................................
..............X.
!!    1 - neg/abstract-explaintypes.scala           [output differs]
!!    2 - neg/warn-unused-params.scala              [output differs]
# 1025/1027 passed, 2 failed in neg
```

Well, it's tidy but it looks like there is an off-by-one race condition when outputting dots, three lines are too long. It's the correct total number of dots, but each line should be 72 dots.

**Comment by adriaanm** [comment]
侘寂

**Comment by som-snytt** [comment]
I don't see my stack trace:

```
# starting 1 test in neg
?! 1 - neg/t11090.scala                          [caught NullPointerException - null]
##### Log file '/Users/andrew/workspace/scala/test/files/neg/t11090-neg.log' from failed test #####

error: null

# 0/1 passed, 1 failed in neg
```

**Pull request merged**